### PR TITLE
AOTAutograd: Go down inference path if no outputs require grad

### PR DESCRIPTION
**Pull request by bdhirsh**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* (to be filled)

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -4135,6 +4140,19 @@ def convert(idx, x):
                     keep_input_mutations=aot_config.keep_inference_input_mutations and not needs_autograd,
                     is_train=needs_autograd,
                 )(*fake_flat_args)
+                if needs_autograd and not any(x.requires_grad for x in fw_metadata.output_info):
+                    # A bit silly: we realized that none of the outputs require grad,
+                    # so we actually have an inference graph.
+                    # Need to recompute metadata since needs_autograd determines some of our ViewAndMutationMeta
+                    # TODO: we should be able to update run_functionalized_fw_and_collect_metadata
```
Comment (Chillee) [inline torch/_functorch/aot_autograd.py:4147]: I think this will increase compilation time - is it possible to avoid?

Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:4147]: Yeah- very dumb but this only matters for subclasses. I’ll update it to only do this for subclasses and a TODO to clean it up

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -4470,7 +4488,7 @@ def aot_function(
     aot_config = AOTConfig(
         fw_compiler=fw_compiler,
         bw_compiler=bw_compiler,
-        inference_compiler=fw_compiler,
+        inference_compiler=inference_compiler,
```
Comment (Chillee) [inline torch/_functorch/aot_autograd.py:4491]: Lol I guess nobody uses this code path anymore

**Comment by bdhirsh** [comment]
Hmm I'm not sure what happened, but I have two ghstack PR's up (this one and https://github.com/pytorch/pytorch/pull/111011)

I can't `ghstack checkout` this PR locally for some reason - ghstack complains that `origin/gh/bdhirsh/468/orig` doesn't exist. I'll just close this PR and work on landing https://github.com/pytorch/pytorch/pull/111011

**Pull request closed without merging**